main.go

```diff
@@ -1,37 +1,73 @@
 package main
 
 import (
+	"bufio"
 	"fmt"
+	"os"
+	"strings"
 )
 
 type Transaction struct {
 	parent *Transaction
 	data   map[string]string
 }
 
-func NewTransaction(parent *Transaction) *Transaction {
+func NewTransaction(p *Transaction) *Transaction {
 	m := make(map[string]string)
-	return &Transaction{data: m}
+	return &Transaction{parent: p, data: m}
+}
+
+func Start(head *Transaction) *Transaction {
+	// parent is current
+	n := NewTransaction(head)
+	// copy map data
+	for k, v := range head.data {
+		n.data[k] = v
+	}
+	return n
 }
 
 func (n *Transaction) Write(k, v string) {
 	n.data[k] = v
 }
 
-func (t *Transaction) Read(k string) string {
+func (t *Transaction) Read(k string) (string, error) {
 	if v, ok := t.data[k]; ok {
-		return v
+		return v, nil
 	}
-	return "Key not found: " + k
+	return "", fmt.Errorf("Key not found: %s", k)
 }
 
 func main() {
-	// initialize with new no parent
+	// initialize with no parent
 	head := NewTransaction(nil)
+	fmt.Println(head) // to see if it's working...
 
-	fmt.Println(len(head.data))
-	head.Write("cat", "cat")
-	fmt.Println(len(head.data))
-	fmt.Println(head.Read("cat"))
-	fmt.Println(head.Read("bat"))
+	scanner := bufio.NewScanner(os.Stdin)
+	for scanner.Scan() {
+		input := strings.Split(scanner.Text(), " ")
+		if input[0] == "quit" {
+			break
+		} else {
+			switch strings.ToUpper(input[0]) {
+			case "START":
+				fmt.Println("Start called")
+				head = Start(head) //to see if it's working...
+				fmt.Println(head)  //to see if it's working...
+			case "WRITE":
+				head.Write(input[1], input[2])
```

Comment:
what happens if the user specifies a command of just "write" with no further arguments?